src/encoder.rs

```diff
@@ -1868,6 +1867,33 @@ fn encode_partition_topdown(fi: &FrameInvariants, fs: &mut FrameState,
     }
 }
 
+use rayon::prelude::*;
+
+
+#[inline(always)]
+fn build_coarse_pmvs(fi: &FrameInvariants, fs: &FrameState) -> Vec<[Option<MotionVector>; REF_FRAMES]> {
+// initial coarse ME loop
+  assert!(!fi.sequence.use_128x128_superblock);
+  let sby_range = 0..fi.sb_height;
+  let sbx_range = 0..fi.sb_width;
+
+  let sbos = (sby_range).flat_map(|sby| {
+      sbx_range.clone().map(move |sbx| SuperBlockOffset { x: sbx, y: sby })
+  }).collect::<Vec<SuperBlockOffset>>();
+
+  sbos.par_iter().map(|sbo| {
+      let bo = sbo.block_offset(0, 0);
+      let mut pmvs: [Option<MotionVector>; REF_FRAMES] = [None; REF_FRAMES];
+      for i in 0..INTER_REFS_PER_FRAME {
+        let r = fi.ref_frames[i] as usize;
+        if pmvs[r].is_none() {
+          pmvs[r] = estimate_motion_ss4(fi, fs, BlockSize::BLOCK_64X64, r, &bo);
```

Comment:
The estimate_motion_ss4() and estimate_motion_ss2() access 
  pub input_hres: Plane,
  pub input_qres: Plane,
fields in "pub struct FrameState { }", then don't they need be Arc? Sinced they are shared resource by threads?

Reply:
No, they are part of the struct and the struct is immutable.

The `Arc<Frame>` is used when you need to share multiple references across different structures and threads.

Reply:
I asked this because of lifetime, not because of mutability.

I just learned nice lesson this morning that Arc< > is needed when caller passes immutable resource to threads where it is shared by them and the lifetime of the resource will not end until all threads are done with the resource. (And clone of Arc<> type storage actually not copying it but generate Arc smart pointer.) 

"Sharing immutable data across threads"
goo.gl/HspHX8 (page 464, 465)

Does rust know that 'pub input_qres' , which is shared by multiple threads of estimate_motion_ss4(), has lifetime until all threads are over?